File: backend/src/main/java/br/ufpr/sept/so2/modules/academico/api/dto/CursoResponse.java
package br.ufpr.sept.so2.modules.academico.api.dto;

import br.ufpr.sept.so2.modules.academico.domain.Curso;
import com.fasterxml.jackson.annotation.JsonProperty;

import java.time.OffsetDateTime;
import java.util.Map;
import java.util.UUID;

public record CursoResponse(
        UUID id,
        String nome,
        String sigla,
        String codigo,
        UUID idCoordenador,
        int horasFormativasMinimas,
        boolean ativo,
        OffsetDateTime createdAt,
        OffsetDateTime updatedAt,
        @JsonProperty("_links") Map<String, String> links
) {

    public static CursoResponse from(Curso curso) {
        String base = "/academico/cursos/" + curso.getId();
        return new CursoResponse(
                curso.getId(),
                curso.getNome(),
                curso.getSigla(),
                curso.getCodigo(),
                curso.getIdCoordenador(),
                curso.getHorasFormativasMinimas(),
                curso.isAtivo(),
                curso.getCreatedAt(),
                curso.getUpdatedAt(),
                Map.of(
                        "self", base,
                        "atualizar", base,
                        "excluir", base,
                        "disciplinas", base + "/disciplinas"
                )
        );
    }
}

File: backend/src/main/kotlin/br/ufpr/sept/so2/modules/academico/api/dto/CursoResponse.kt
package br.ufpr.sept.so2.modules.academico.api.dto

import br.ufpr.sept.so2.modules.academico.domain.Curso
import com.fasterxml.jackson.annotation.JsonProperty
import java.time.OffsetDateTime
import java.util.UUID

data class CursoResponse(
    val id: UUID,
    val nome: String,
    val sigla: String,
    val codigo: String,
    val idCoordenador: UUID?,
    val secretariosIds: List<UUID>,
    val horasFormativasMinimas: Int,
    val ativo: Boolean,
    val createdAt: OffsetDateTime,
    val updatedAt: OffsetDateTime,
    @get:JsonProperty("_links")
    val links: Map<String, String>,
) {
    companion object {
        fun from(
            curso: Curso,
            secretariosIds: List<UUID>,
            podeGerenciar: Boolean,
            podeVerDisciplinas: Boolean,
        ): CursoResponse {
            val base = "/academico/cursos/${curso.id}"
            val links = linkedMapOf("self" to base)
            if (podeGerenciar) {
                links["atualizar"] = base
                links["excluir"] = base
            }
            if (podeVerDisciplinas) {
                links["disciplinas"] = "/academico/disciplinas?idCurso=${curso.id}"
            }
            return CursoResponse(
                id = curso.id,
                nome = curso.nome,
                sigla = curso.sigla,
                codigo = curso.codigo,
                idCoordenador = curso.idCoordenador,
                secretariosIds = secretariosIds,
                horasFormativasMinimas = curso.horasFormativasMinimas,
                ativo = curso.ativo,
                createdAt = curso.createdAt,
                updatedAt = curso.updatedAt,
                links = links,
            )
        }
    }
}
